[cmd/opampsupervisor] Implement startup fallback configuration

douglascamata opened the pull request:
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->
#### Description

This PR adds a new feature for the Supervisor which allows it to start the Collector with an "startup fallback" configuration in case it cannot connect to the OpAMP backend.

When the connection to the OpAMP backend is restored, this configuration should be abandoned by the Supervisor in favor of the files already in `agent::config_files`.

The startup configuration fallback feature is only enabled when at least 1 or more configuration files are passed via the `agent::startup_fallback_configs` config option. When 2 or more files are provided, they are merged in the order in which they are provided.

By default this feature is off (i.e. `agent::startup_fallback_configs` is not provided).

<!-- Issue number (e.g. #1234) or full URL to issue, if applicable. -->
#### Link to tracking issue
Fixes #44368. 

<!--Describe what testing was performed and which tests were added.-->
#### Testing

<!--Describe the documentation added.-->
#### Documentation

<!--Please delete paragraphs that you did not use before submitting.-->

douglascamata committed:
Implement fallback configuration

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata committed:
Fix formatting

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata committed:
Add changelog entry

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata committed:
Fix tests

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata committed:
Merge branch 'main' into supervisor-fallback-config

douglascamata committed:
Merge branch 'main' into supervisor-fallback-config

douglascamata commented on cmd/opampsupervisor/supervisor/supervisor.go:
```diff
@@ -1472,6 +1499,165 @@ func (s *Supervisor) composeMergedConfig(incomingConfig *protobufs.AgentRemoteCo
+		return false, fmt.Errorf("could not read fallback config file: %w", err)
+	}
+
+	if err = k.Load(rawbytes.Provider(fallbackConfigBytes), yaml.Parser(), koanf.WithMergeFunc(configMergeFunc)); err != nil {
```
That's what the `opampsupervisor` package uses already for loading/processing configuration, @atoulme. I don't want to introduce too many new things here. I believe eventually it should converge to use [open-telemetry/opentelemetry-collector@main/confmap](https://github.com/open-telemetry/opentelemetry-collector/tree/main/confmap?rgh-link-date=2025-12-31T23%3A05%3A06Z), but that'll be a sizeable refactor.

atoulme replied:
understood

atoulme commented on cmd/opampsupervisor/README.md:
```diff
@@ -129,6 +129,48 @@ For more details on the healthcheck configuration, see the see the [full list of
+
+- The fallback configuration is a **standalone configuration file**. It is intentionally not merged with the `config_files` setting. This ensures predictable fallback behavior without dependencies on other configuration files.
+
+- The OpAMP extension configuration is automatically added to the fallback configuration to maintain communication between the Supervisor and the Collector.
```
telemetry as well

douglascamata committed:
Merge branch 'main' into supervisor-fallback-config

douglascamata committed:
Remove useless comment

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata committed:
Add test for the fallback config composition

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata committed:
Improve description of the fallback config

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata commented on cmd/opampsupervisor/supervisor/supervisor.go:
```diff
@@ -1472,6 +1499,165 @@ func (s *Supervisor) composeMergedConfig(incomingConfig *protobufs.AgentRemoteCo
+		}
+	}
+
+	// The merged final result is our new merged config.
```
Test added in [9f2f619](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/45100/commits/9f2f6192d04f6013a4a9314d9b158a660fb5c999)

douglascamata committed:
Retrigger build

douglascamata commented:
@atoulme my general idea with this is to allow customers of Coralogix (or any other vendor) to create a configuration that they think is essential for their Collectors that are managed remotely via OpAMP in a way that outages of the OpAMP backend are not bringing down their observability. Outages of the OpAMP backend are capable of bringing down a part of your fleet of Supervised Collectors, because the Supervisor will refuse to bring up a Collector if it can't talk to the backend.

I don't think it makes sense for the Otel Supervisor release to provide a specific fallback configuration included by default. The usability of this configuration is on the fact that it heavily tailored to each user as they have their own pipelines with their own components and only they know what's important there. Any 3rd party cannot judge what's critically important for a user in their Collector configuration.

douglascamata commented:
What I am doing here is providing just the fallback feature. I do not provide any default fallback configuration within the `opampsupervisor` package.

atoulme reviewed (dismissed):
I think this is a powerful feature. I feel this is an interesting increment in the capabilities of the opampsupervisor. I am weary of the feature set increasing piece by piece like this and we might need to have a roadmap discussion and a bit of thought in how we intend to manage this tool's use cases moving forward.

I would like at least one review from either @tigrannajaryan or @evan-bradley as codeowners to make sure we are engaging in the right direction.

evan-bradley commented on cmd/opampsupervisor/supervisor/connstate.go:
```diff
@@ -0,0 +1,282 @@
+	prevState := t.state
+	t.state = ConnectionStateConnected
+
+	t.logger.Info("OpAMP connection established",
```
Aren't we already logging this somewhere else? Can we consolidate the logs if we are? I want to keep info logs to a minimum if possible to ensure a high signal-to-noise ratio at that level of logging.

evan-bradley reviewed (commented):
Thanks for taking this on. I'm still working through it, but won't be able to finish right away, so want to provide what feedback I have so far.

douglascamata commented on cmd/opampsupervisor/README.md:
```diff
@@ -129,6 +129,48 @@ For more details on the healthcheck configuration, see the see the [full list of
+
+| Option | Description |
+|--------|-------------|
+| `fallback_config` | Path to a fallback configuration file to use when the OpAMP server is unreachable. This must be a complete, standalone Collector configuration. |
```
I'm not sure about the use case of having the configuration split into multiple files vs a single big file, in general. Personally I'm yet to see someone using multiple config files, but I'm +1 for supporting it as well. Will update the implementation.

douglascamata commented on cmd/opampsupervisor/supervisor/config/config.go:
```diff
@@ -264,9 +275,41 @@ func (a Agent) Validate() error {
+	}
+
+	// Validate that the fallback config file exists
+	if _, err := os.Stat(a.FallbackConfig); err != nil {
```
Good idea. Will do it.

douglascamata committed:
Remove useless logs

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata committed:
Improve log msg

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata committed:
Use 30s as default fallback startup timeout

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata committed:
Remove useless comments

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata committed:
Validate fallback config using the agent executable

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata committed:
Add default fallback startup timeout to supervisor config test

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata committed:
Refactor fallback config from a singel config to a list

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata committed:
Update and improve supervisor's READMEs

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata committed:
Default fallback configs should be `nil`

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata committed:
Retrigger build

evan-bradley commented on .chloggen/opampsupervisor-fallback-config.yaml:
```diff
@@ -0,0 +1,33 @@
+  2. Lost connection to the OpAMP backend after startup, controlled by `fallback_runtime_timeout`.
+  Both options are time durations, allowing the OpAMP client to handle retries before switching.
+  They can be disabled by setting them to zero or omitting them entirely. The two timeouts are
+  independent, allowing different behaviors for startup vs runtime connection failures.
```
Thanks for the discussion, sounds like we're in agreement. Let's remove it for now and we can add it back once we have more concrete uses in mind.

douglascamata committed:
Completely remove the runtiem configuration fallback feature

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata committed:
`go fmt`

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

evan-bradley commented:
@douglascamata sorry for the delay. I have an in-progress review, but want to make sure I give this PR the attention it deserves, and have been unable to find the time.

I'm really sorry to do this so far into the review process, but I also want to suggest that we temporarily pause this. @atoulme expressed some concerns at adding this functionality at today's SIG meeting (2026-01-21), and since we don't have any documented vision for the Supervisor, I want to take a moment to step back and ensure everyone agrees on adding this functionality. The Supervisor is intended to be highly-reliable, and any code we add increases its complexity and introduces risk for reliability issues.

I would still like to see this added, as I think it makes sense to offer the ability to start the Collector with a given config in the case where the Supervisor is unable to connect to an OpAMP server. I will get you an update by this Friday at the latest.

douglascamata committed:
Merge branch 'main' into supervisor-fallback-config

evan-bradley reviewed (commented):
Thanks for your patience. I still think this makes sense for increasing reliability for supervised Collectors, I've left some high-level comments around the general approach. I will reach out on the #otel-opamp Slack channel to confirm with others and point them here if they have specific feedback.

douglascamata committed:
Update README explanation about fallback configuration

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata committed:
Merge branch 'supervisor-fallback-config' of github.com:douglascamata/opentelemetry-collector-contrib into supervisor-fallback-config

douglascamata committed:
Simplify fallback config logic

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata committed:
Remove all mentions to `fallback_startup_timeout`

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata committed:
Extract windows path fix to a function

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata committed:
Merge branch 'main' into supervisor-fallback-config

douglascamata committed:
Retrigger build

douglascamata committed:
Remove new config signal from `loadAndWriteInitialMergedConfig`

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata committed:
Signal new config when agent ready

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata committed:
Improve var names and if logic order

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata committed:
Remove check for persisted remote config presence

The check for config state is already enough, as it checks for the
presence fo the merged configuration. This will already include any
remote configuration that might have been received.

Signed-off-by: Douglas Camata
<159076+douglascamata@users.noreply.github.com>

douglascamata committed:
Remove unused function

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata committed:
Retrigger build

douglascamata committed:
Revert "Remove unused function"

This reverts commit f55a26b6b543cd8bad57065ca7fb56e3b8fcbb73.

douglascamata committed:
Revert "Remove check for persisted remote config presence"

This reverts commit 27c818f8e2198b4a1b03d5e5b95253804291d69a.

douglascamata committed:
Remove unnessary changes from e2e tests

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata committed:
Fix hup reload never setting health when agent doesn't become ready

This started to happen in e2e tests in CI, which I guess is slower
because of shared ifnra. I could not reproduce it locally until I tried
bumping the bootstrap timeout to 10s. With this, it can be reliably
reproduced and this fix makes it work.

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata committed:
Readd code removed by accident

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata committed:
Fix lint

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata committed:
Update changelog

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata committed:
Rename "fallback config" to "initial fallback config"

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata committed:
Merge branch 'main' into supervisor-fallback-config

douglascamata committed:
Synchronize writes to initial config

It could be written by the initiatilization logic and the opamp server
callbacks, so it requires some synchronization to ensure there is no
data race.

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata committed:
Update comment

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata committed:
Merge branch 'main' into supervisor-fallback-config

douglascamata commented:
Any chance of getting a review? 🥺

douglascamata committed:
Merge branch 'main' into supervisor-fallback-config

evan-bradley commented on .chloggen/opampsupervisor-fallback-config.yaml:
```diff
@@ -0,0 +1,32 @@
+component: cmd/opampsupervisor
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Start the Collector with an initial fallback config until successfully connected to the OpAMP backend
```
```suggestion
note: Allow starting the Collector with an initial fallback config until successfully connected to the OpAMP backend
```
We should make it clear this isn't default behavior.

douglascamata replied:
Applied this as part of [62fcbbb](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/45100/commits/62fcbbbc077b52c5619563794c2963f4a3e4021e)

evan-bradley commented on cmd/opampsupervisor/examples/supervisor_fallback.yaml:
```diff
@@ -0,0 +1,37 @@
+  # Fallback configuration file to use when OpAMP server is unreachable.
+  # This must be a complete, standalone Collector configuration.
+  # Note: This is NOT merged with config_files - it's used as-is when fallback is triggered.
+  initial_fallback_configs:
```
I know we've gone back and forth on the name a few times, so thanks for your patience with this. Now that I'm coming back to this with fresh eyes, what do you think about just `fallback_configs`? We can specify in the docs that fallback only happens before the Supervisor connects to the OpAMP server.

Alternatively, I would vote for `startup_fallback_configs. I think that makes it clearer these are used on Supervisor startup, as where "initial" implies to me that there are other fallback configs that come later.

douglascamata replied:
We can rename it, no worries. In fact, I really like your proposal of `startup_fallback_configs`: it's clear, concise and will confuse less people. I can still specify in docs that it only applies at start-up, to be sure that nobody will misunderstand it.

douglascamata replied:
Fixed in [62fcbbb](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/45100/commits/62fcbbbc077b52c5619563794c2963f4a3e4021e)

evan-bradley commented on cmd/opampsupervisor/supervisor/supervisor.go:
```diff
@@ -1485,6 +1536,95 @@ func (s *Supervisor) composeMergedConfig(incomingConfig *protobufs.AgentRemoteCo
+		}
+	}
+
+	// Add the OpAMP extension config
```
Could we add a method that adds all the extra config layers so this code is shared with the main config compose method?

douglascamata replied:
Sure. I'll refactor it a bit.

douglascamata replied:
Fixed in [78aecff](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/45100/commits/78aecff031efaa17cfa868470592e3432d1ba160)

evan-bradley reviewed (commented):
Overall looks good to me.

We should probably have a follow-up to allow a timeout before fallback configs are used.

pitr commented:
would it also make sense to add a new metric to [telemetry/metrics.go](https://github.com/douglascamata/opentelemetry-collector-contrib/blob/main/cmd/opampsupervisor/supervisor/telemetry/metrics.go), something akin to `supervisor.agent.health_status`? Maybe `supervisor.agent.fallback_status`

douglascamata commented:
> would it also make sense to add a new metric to [telemetry/metrics.go](https://github.com/douglascamata/opentelemetry-collector-contrib/blob/main/cmd/opampsupervisor/supervisor/telemetry/metrics.go?rgh-link-date=2026-04-15T09%3A10%3A19Z), something akin to `supervisor.agent.health_status`? Maybe `supervisor.agent.fallback_status`

Sounds like a great idea, @pitr. I will try to include it it if the change is small enough, I don't foresee a reason for it not to be.

douglascamata committed:
Rename `initial_fallback_configs` -> `startup_fallback_configs`

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata committed:
Extract method to load extra configs for fallback

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata committed:
Add metric to track whether fallback config is being used

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata commented:
@pitr added the metric in [8f16c9d](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/45100/commits/8f16c9d7a5fe4f2985bbf4269fe0043e2ff8c1e3)

douglascamata committed:
Make linter happy

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata committed:
Linter fix, again

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

cforce commented:
>This PR adds a new feature for the Supervisor which allows it to start the Collector with an "initial fallback" configuration in >case it cannot connect to the OpAMP backend.

Isn't this  alreay covered by custom last line of cfg ?

# List of configuration files to be merged to build the Collector's effective
  # configuration. It includes a few "special" files. Read the "Config Files" section
  # below for more details.
  config_files:
    - $OPAMP_EXTENSION_CONFIG
    - $OWN_TELEMETRY_CONFIG
    - $REMOTE_CONFIG
    - myLocylFallBackCfg.yaml

douglascamata commented:
> Isn't this alreay covered by custom last line of cfg ?
> (...)
> config_files:
> - $OPAMP_EXTENSION_CONFIG
> - $OWN_TELEMETRY_CONFIG
> - $REMOTE_CONFIG
> - myLocylFallBackCfg.yaml

@cforce it's not, because of two things:

1. If the Supervisor cannot talk to the OpAMP backend at startup it won't even start a Collector for you. 
2. In your example, your fallback configuration is **always** merged with the other configuration files, which might not be always what the user wants.

Users might want to have a completely separate configuration for this. For example, if the OpAMP backend has the capability of exporting your remote configurations to a different storage mechanism (i.e. S3) or if you manage you remote configs via gitops, you can bring them into the container (there are many possibilities for this) and load them until the connection to the backend is restored. 

Other subset of users might be using an OpAMP backend without these capabilities and they can still rely on having a local file with minimal configuration so that they don't lose the capability of starting new Collectors **with some useful configuration** until they can reach the backend.

tl;dr: having a separate configuration entry for startup fallback configuration provides more flexibility and allows many different scenarios to be handled as users see fit.

douglascamata committed:
Merge branch 'main' into supervisor-fallback-config

cforce commented:
1. Afaik this Is not true - collector is started based on last successful received cfg -also when rhere is no connectivity
2. when yöu fallback -then because there is no prior cfg from cloud(empty) and the fallback.yaml would apply only. If there is already is s opamp backend downloaded cfg it will merge it with local

douglascamata commented:
> 1. Afaik this Is not true - collector is started based on last successful received cfg -also when rhere is no connectivity
> 2. when yöu fallback -then because there is no prior cfg from cloud(empty) and the fallback.yaml would apply only. If there is already is s opamp backend downloaded cfg it will merge it with local

@cforce: The information you have there is not the whole picture. The Collector is only started **if there is a last effective configuration stored in disk**. If you have a k8s cluster that auto scales nodes up and down, for example, new Collector pods won't start because the OpAMP backend is down and there's no last effective configuration in disk. 

This is why the feature is called **startup** fallback configuration: it's only useful at the first startup of the Collector.

douglascamata committed:
Replace `initial` -> `startup` in changelog entry

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata commented:
@cforce what I mentioned in my previous comment is explained in the changelog entry related to these changes:

> **When there's no previous configuration state persisted in disk**, the Supervisor now
> starts with a startup fallback configuration until it can successfully connect
> to the OpAMP backend. At this moment the regular configuration (indicated by
> `agent::config_files`) is restored.
> **If a previous configuration state is persisted**, it will be used to start the Collector.

evan-bradley commented on cmd/opampsupervisor/README.md:
```diff
@@ -129,6 +129,38 @@ For more details on the healthcheck configuration, see the see the [full list of
+
+| Option | Description |
+|--------|-------------|
+| `startup_fallback_configs` | List of paths to startup fallback configuration files to use. If more than one path is specified, they are merged in order. Together, these must be complete, standalone Collector configuration. |
```
```suggestion
| `startup_fallback_configs` | List of paths to startup fallback configuration files to use. If more than one path is specified, they are merged in order. Together, these must form a complete, standalone Collector configuration. |
```

evan-bradley commented on cmd/opampsupervisor/README.md:
```diff
@@ -129,6 +129,38 @@ For more details on the healthcheck configuration, see the see the [full list of
+
+The Supervisor supports a startup fallback configuration mechanism that provides resilience when the OpAMP server is unreachable at startup and there's no previous configuration state persisted in disk. This is useful for ensuring the Collector can start with a known-good configuration during network outages or server maintenance. When the Supervisor successfully connects to the OpAMP server, the ...
+
+If when the Supervisor starts there's previous configuration state persisted in disk, the Supervisor will use it to start the Collector.
```
```suggestion
If, when the Supervisor starts, there's previous configuration state persisted in disk, the Supervisor will use it to start the Collector.
```

evan-bradley reviewed (commented):
Overall looks good. Thank you for being so patient with reviews on this.

There's a conflict in the E2E test, if you can fix that, I'll do a quick final pass tomorrow.

douglascamata committed:
Merge branch 'main' into supervisor-fallback-config

douglascamata committed:
Fix some grammar

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata committed:
Fix tests after merge

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

evan-bradley reviewed (approved):
Thanks for your persistence throughout the review process for this. It looks good to me, but having a second review would be nice if possible. If we don't get one by next Wednesday (2026-05-20), I'll merge this.

douglascamata commented:
Sounds good to me, @evan-bradley. Reviews don't cost nothing, so thank you for the time as well. 😄

dpaasman00 commented on cmd/opampsupervisor/supervisor/supervisor.go:
```diff
@@ -635,10 +647,34 @@ func (s *Supervisor) startOpAMPClient() error {
+				s.telemetrySettings.Logger.Info("Connected to the OpAMP server.")
+
+				s.configWriteMu.Lock()
+				defer s.configWriteMu.Unlock()
```
We don't need to claim this lock until after the `fallbackEnabled` check, right?

dpaasman00 commented on cmd/opampsupervisor/supervisor/supervisor.go:
```diff
@@ -635,10 +647,34 @@ func (s *Supervisor) startOpAMPClient() error {
+				if err := s.waitForAgentReady(); err != nil {
+					s.telemetrySettings.Logger.Debug("Agent not ready before config update; skipping signal", zap.Error(err))
+					return
+				}
```
Do we need to call `waitForAgentReady` here? It looks like the only other place it's called is in `hupReloadAgent` which will be called when the event is pulled off the `s.hasNewConfig` channel.

douglascamata replied:
@dpaasman00: I was hitting a data race without this on the access to `s.config.Agent.ConfigFiles` because `Start` kicks off the OpAMP client loop and calls `loadAndWriteInitialMergedConfig`. 

The new code on the `OnConnect` function was racing to copy the whole `s.config.Agent` (because of a value-receiver method) that `loadAndWriteInitialMergedConfig` was modifying (because it changes & reassigns `s.config.Agent.ConfigFiles`). 🤯 

I changed how this works by using a copy of the config files in `s.config.Agent.ConfigFiles` instead of modifying them in-place.

dpaasman00 commented on cmd/opampsupervisor/supervisor/supervisor.go:
```diff
@@ -635,10 +647,34 @@ func (s *Supervisor) startOpAMPClient() error {
 		RemoteConfigStatus: s.persistentState.GetLastRemoteConfigStatus(),
 		Callbacks: types.Callbacks{
 			OnConnect: func(context.Context) {
```
More of a style thing, but are we able to refactor `OnConnect` into a separate function that's called here instead of being defined inline?

dpaasman00 commented on cmd/opampsupervisor/supervisor/supervisor.go:
```diff
@@ -635,10 +647,34 @@ func (s *Supervisor) startOpAMPClient() error {
+				if !opampBackendConnected {
+					return
+				}
+				s.telemetrySettings.Logger.Info("Connected to the OpAMP server.")
```
I think this log should still be the first line in this function. In the event the supervisor's connection to the server is dropped and then re-established, the check at the start will return before we log this.

dpaasman00 commented on cmd/opampsupervisor/supervisor/supervisor.go:
```diff
@@ -635,10 +647,34 @@ func (s *Supervisor) startOpAMPClient() error {
 		Callbacks: types.Callbacks{
 			OnConnect: func(context.Context) {
-				s.telemetrySettings.Logger.Info("Connected to the server.")
+				opampBackendConnected := s.initialOpampConnSuccess.CompareAndSwap(false, true)
```
Thoughts on renaming this var to `firstTimeConnected` for readability?

dpaasman00 reviewed (changes requested):
Some initial thoughts, looks good though. Will test next.

Is it at all possible a race condition exists such that a remote config message from the server could get effectively dropped at connection time because the supervisor is busy starting the collector with the fallback config, then pivoting to the standard config, then pivoting again to the remote config? It's possible that the supervisor runs the collector with 3 different configurations in a short window at start up.

douglascamata committed:
Apply review feedback

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata commented:
@dpaasman00 thanks for the review. I took care of all your comments & suggestions, so please have another look whenever you can.

dpaasman00 commented on cmd/opampsupervisor/supervisor/supervisor.go:
```diff
@@ -968,6 +978,30 @@ func (*Supervisor) getHeadersFromSettings(protoHeaders *protobufs.Headers) http.
 
+func (s *Supervisor) onConnect(_ context.Context) {
+	s.telemetrySettings.Logger.Info("Connected to the OpAMP server.")
+	opampBackendConnected := s.firstTimeConnected.CompareAndSwap(false, true)
```
Sorry, I meant changing the `opampBackendConnected` var to `firstTimeConnected`, the atomic.Bool can remain the previous value. Doing the mental conversion of when this is true vs false based on the CompareAndSwap was making the following conditional a bit hard to read.

douglascamata replied:
@dpaasman00 no worries. To be honest, I also find `CompareAndSwap` not easy to read and understand. After I fixed the code to incorporate your feedback it struck me that it might be better to make this easier to read even if not as optimal, so I did this in [dea9966](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/45100/commits/dea99668fa6237ab432d72d8f3379daaf076753b)

dpaasman00 reviewed (approved):
Working well for me!

douglascamata committed:
Fix PR feedback

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata committed:
Remove intermediate var to simplify code

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata committed:
Retrigger build

douglascamata committed:
Fix logic around simplified code

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata committed:
Merge branch 'main' into supervisor-fallback-config

douglascamata committed:
Merge branch 'main' into supervisor-fallback-config

douglascamata committed:
Fix fallback config validation tests

Now `config.Load` doesn't validate the configurations, so we need to
create a new Supervisor to validate it.

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata commented:
Had to do some other fixes related to a broken main build and the changes coming from the Supervisor extensions (config validation logic moved). But this is good for another review when you can, @dpaasman00

dpaasman00 commented on cmd/opampsupervisor/supervisor/supervisor.go:
```diff
@@ -987,6 +997,31 @@ func (*Supervisor) getHeadersFromSettings(protoHeaders *protobufs.Headers) http.
 	return headers
 }
 
+func (s *Supervisor) onConnect(_ context.Context) {
```
This function should call `SetCollectorFallbackStatus` with a `false` value, right? Since this is when we reset the configuration in the event we were running a fallback config and have now connected to the server?

douglascamata replied:
Good catch. I didn't have a test covering this, so I add one together with the fix. 👍

dpaasman00 commented on cmd/opampsupervisor/supervisor/config/config.go:
```diff
@@ -275,9 +279,56 @@ func (a Agent) Validate() error {
+	return nil
+}
+
+func (a Agent) validateFallbackConfigsWithColBin() error {
```
I understand wanting to validate these configuration files before running them, but I'm not convinced doing so in this area of the code is the right place. 

Thoughts on moving the logic that runs the collector's validation of the configs into the `composeFallbackConfig()` function? This would more closely match how the supervisor validates `agent::config_files`.

douglascamata replied:
@dpaasman00 I validate them before running to be sure that they will work, and by this I mean that at least the Collector won't immediately crash, when I need them. `composeFallbackConfig` only runs when it's time to use the fallback config. This works fine for `agent::config_files` because they are composed immediately at startup.

If only when I need the fallback config I discover that it crashes the Collector, it defeats the whole purpose of the feature (a config to keep my Collector working when it has no config from the OpAMP backend) and leaves me with without observability.

dpaasman00 reviewed (changes requested):
2 comments:
- The comment about the metric is a requested change as I think this is where we should be setting the metric
- The config validation comment is more of a question/statement. Genuinely curious what your thoughts are.

douglascamata committed:
Ensure fallback status metric is false on connect

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata committed:
Merge branch 'main' into supervisor-fallback-config

evan-bradley reviewed (approved):
Thanks @douglascamata for persisting through a long review cycle.

Also really appreciate your in-depth review and manual testing @dpaasman00.
